### ISSUE TITLE
Using xgboost with irace

### ISSUE THREAD

**Issue by JannesRausch** [mlr-org/mlr#1454]
Hello I am an R beginner and I have a question regarding the xgboost.
I´m trying to make a hyperparametertuning. I try to use Irace from the mlr package to get a great Model for my Classification problem. At first i tried irace with a Randomforrest Model and it works great. But know I need a xgboost Model, but it does not work. I use the "UCI Bankmarketing Data Set" it has 17 Variables (Nummerics, Integer and Char) with the Target Variable "y" ( y/n).

`
Here my Code:
library(xgboost)
library(mlr)

#dataset 
bmdata=read.table(file.choose(),header=TRUE,sep=",")

#task and learner
task = makeClassifTask(data = bmdata, target = "y")
lrn = makeLearner("classif.xgboost", predict.type = "prob", fix.factors.prediction = TRUE)

#parameterset
ps = makeParamSet(
  makeDiscreteParam("booster", values=c("gbtree","gblinear")),
  makeIntegerParam("silent",lower=-0, upper=1L),
  makeNumericParam("eta",lower=0, upper=1),
  makeNumericParam("min_child_weight",lower=0, upper=1),
  makeNumericParam("subsample",lower=0, upper=1),
  makeNumericParam("colsample_bytree",lower=0, upper=1),
  makeIntegerParam("num_parallel_tree",lower=1, upper=30),
  makeNumericParam("lambda",lower=0, upper=1),
  makeNumericParam("lambda_bias",lower=0, upper=1),
  makeNumericParam("alpha",lower=0, upper=1),
  makeNumericParam("base_score",lower=-5, upper=5), 
  makeNumericParam("missing",lower=-5, upper=5),
  makeIntegerParam("nthread",lower=1L, upper=15),
  makeIntegerParam("nrounds",lower=1, upper=5),
  makeIntegerParam("verbose",lower=0, upper=2),
  makeIntegerParam("print.every.n",lower=1, upper=2),
  makeIntegerParam("early.stop.round",lower=1, upper=2),
  makeLogicalParam(id = "maximize")
)

#resample and irace
rdesc=makeResampleDesc(method="Holdout")
ctrl = makeTuneControlIrace(maxExperiments = 1200)
finetune = tuneParams(lrn, task = task, resampling = rdesc, par.set = ps, control = ctrl, measures = auc)
finetune

#We can generate a Learner with optimal hyperparameter settings as follows:
optlrn  = setHyperPars(makeLearner(" classif.xgboost", predict.type = "prob"), par.vals = finetune$x)
optlrn
optmod = train(optlrn,task,subset= DATA_IMPORT)
`
The data for the train come from SAS and the Prediction will be executed in SAS

I get the error 
Error in xgb.DMatrix(data, label = label, missing = missing) : 
  REAL() can only be applied to a 'numeric', not a 'integer'

I tried it with various possibilities I prefer the problem is that I use a data.table not a data.frame or a matrix. 

So i tried it with a Matrix (bmmatrix=data.matrix(bmdata)) When i try to create a task 
"Error in makeClassifTask(data = bmmatrix, target = "y") : 
  Assertion on 'data' failed: Must be of type 'data.frame', not 'matrix'"

I think  I need a Data.Frame but I have no idea how I can convert my table.data to a data.frame.
Am I right with my guess? Or what do I need to make it work?

**Comment by berndbischl** [comment]
can you please provide an fully executable example so we can run this?
you can simply use fake data with the same technical structure

**Comment by JannesRausch** [comment]
`library(xgboost)
library(mlr)

### dataset (bi_ha)
DT<- data.table(Age=20:70,JOB=c("Management","Martial","Admin"), Housing=c("yes","no"),Balance=-2000:2000,Y=c("y","n"))
cols <- c('JOB', 'Housing','Y')
DT[, (cols) := lapply(.SD, factor), .SDcols=cols]

#task and learner
task = makeClassifTask(data = DT, target = "Y")
lrn = makeLearner("classif.xgboost", predict.type = "prob", fix.factors.prediction = TRUE)


#listLearners(task)

#parameterset
ps = makeParamSet(
  makeDiscreteParam("booster", values=c("gbtree","gblinear")),
  makeIntegerParam("silent",lower=-0, upper=1L),
  makeNumericParam("eta",lower=0, upper=1),
  makeNumericParam("min_child_weight",lower=0, upper=1),
  makeNumericParam("subsample",lower=0, upper=1),
  makeNumericParam("colsample_bytree",lower=0, upper=1),
  makeIntegerParam("num_parallel_tree",lower=1, upper=30),
  makeNumericParam("lambda",lower=0, upper=1),
  makeNumericParam("lambda_bias",lower=0, upper=1),
  makeNumericParam("alpha",lower=0, upper=1),
  makeNumericParam("base_score",lower=-5, upper=5), 
  makeNumericParam("missing",lower=-5, upper=5),
  makeIntegerParam("nthread",lower=1L, upper=25),
  makeIntegerParam("nrounds",lower=1, upper=5),
  makeIntegerParam("verbose",lower=0, upper=2),
  makeIntegerParam("print.every.n",lower=1, upper=2),
  makeIntegerParam("early.stop.round",lower=1, upper=2),
  makeLogicalParam(id = "maximize")
)


#resample and irace
rdesc=makeResampleDesc(method="Holdout")
ctrl = makeTuneControlIrace(maxExperiments = 1200)
finetune     = tuneParams(lrn, task = task, resampling = rdesc, par.set = ps, control = ctrl, measures = auc)
finetune

`

**Comment by Stevo15025** [comment]
1. Running your example, finetune gives an error
```r
finetune = tuneParams(lrn, task = task, resampling = rdesc, par.set = ps, control = ctrl, measures = auc)
## Error in checkTunerParset(learner, par.set, measures, control) : 
##  Can only tune parameters for which learner parameters exist: print.every.n,early.stop.round 
```

2. A minimimal reproducible example for your problem is

```r
library(mlr)
library(data.table)
#dataset
DT= data.table(Age=20:70,
                Balance=-2000:2000,
                Y=factor(c("y","n")))
#task and learner
task = makeClassifTask(data = DT, target = "Y")
lrn = makeLearner("classif.xgboost", predict.type = "prob", fix.factors.prediction = TRUE)

trained.mod <- train(lrn, task)
## Error in xgb.DMatrix(data, label = label, missing = missing) : 
##  REAL() can only be applied to a 'numeric', not a 'integer' 
```

From what I see this is intentional on part of the xgboost authors and has nothing to do with mlr. Changing `Age` and `Balance` to numeric runs fine.

```r
DT<- data.table(Age=20:70,
                Balance=-2000:2000,
                Y=factor(c("y","n")))
DT[,`:=`(Age = as.numeric(Age), Balance = as.numeric(Balance))]
#task and learner
task = makeClassifTask(data = DT, target = "Y")
lrn = makeLearner("classif.xgboost",
                  predict.type = "prob",
                  fix.factors.prediction = TRUE)

trained.mod <- train(lrn, task)
```

We could correct for integers that should be numerics in `train.regr.xgboost()`, though idk enough about xgboost to say that is okay.

**Comment by JannesRausch** [comment]
Thanks!! it works... just a little mistake, a beginner's mistake. I always check the Attribute Characteristics with the Rstudio inspect function. And there was standing "numeric Attribute".
Now I checked the  Attributes over "str(DT)" and I saw the that the Attributes are Integers.
Just a formation to Numeric and it works!
Thanks to Stevo15025
